src/types/redis_sb_chain.cc

```diff
@@ -0,0 +1,240 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#include "redis_sb_chain.h"
+
+namespace redis {
+
+void SBChain::appendBFSuffix(const Slice &ns_key, uint16_t filters_index, std::string *output) {
+  output->clear();
+
+  output->append(ns_key.data(), ns_key.size());
+  output->append(kBloomFilterSeparator);
+  PutFixed16(output, filters_index);
+}
+
+void SBChain ::appendBFMetaSuffix(const Slice &bf_key, std::string *output) {
+  output->clear();
+
+  output->append(bf_key.data(), bf_key.size());
+  output->append(kBloomFilterSeparator);
+  output->append("meta");
+}
+
+void SBChain::bfInit(BFMetadata *bf_metadata, std::string *bf_data, uint32_t entries, double error) {
+  bf_metadata->entries = entries;
+  bf_metadata->error = error;
+  bf_metadata->size = 0;
+
+  uint32_t bf_bytes = BlockSplitBloomFilter::OptimalNumOfBytes(entries, error);
+  bf_metadata->bf_bytes = bf_bytes;
+
+  BlockSplitBloomFilter block_split_bloom_filter;
+  block_split_bloom_filter.Init(bf_bytes);
+  *bf_data = block_split_bloom_filter.GetData();
```

Comment:
This would copy these data. I prefer a interface like:

```c++
void SBChain::bfInit(BFMetadata *bf_metadata, uint32_t entries, double error, std::unique_ptr<BlockSplitBloomFilter>) {
  bf_metadata->entries = entries;
  bf_metadata->error = error;
  bf_metadata->size = 0;

  uint32_t bf_bytes = BlockSplitBloomFilter::OptimalNumOfBytes(entries, error);
  bf_metadata->bf_bytes = bf_bytes;

  *block_split_bloom_filter = std::make_unique<BlockSplitBloomFilter>();
  block_split_bloom_filter->Init(bf_bytes);
}
```